sys/kern/tty.c

```diff
@@ -902,6 +962,54 @@ static int tty_ioctl(file_t *f, u_long cmd, void *data) {
   }
 }
 
+static void tty_hangup(tty_t *tty) {
+  assert(mtx_owned(&tty->t_lock));
+
+  if (!tty_opened(tty) || (tty->t_lflag & CLOCAL))
+    return;
+
+  session_t *s = tty->t_session;
+
+  /* Send SIGHUP to session leader.
+   * NOTE: access to s_leader is safe here,
+   * even though we're not holding all_proc_mtx. See session_leave(). */
+  if (s && s->s_leader) {
+    proc_t *leader = s->s_leader;
```

Comment:
I don't get it. Why `s->s_leader` cannot became `NULL` between this lines? Does only session leader can call this function? If not i think it is possible that this function is triggered by some process and `session_leave()` is called in session leader in the same time.

Reply:
Okay, i missed one change. We are holding `tty->t_lock` here and in `session_leave()`. Now I see it.